### ISSUE TITLE
Invalid `Mockery::mock()`/`spy()` phpdoc (= phpstan errors)

### ISSUE THREAD

**Issue by LastDragon-ru** [padraic/mockery#1400]
### Mockery Version

1.6.10

### PHP Version

PHP 8.3

### Issue Description

Today I've updated to the latest 1.6.10 and see a lot of phpstan errors... Seems they were introduced in #1397 where the type of `$args` was changed to `@param array<class-string<TMock>|TMock> $args`. After this change, any of

```
$mock = \Mockery::mock('MyClass', [$constructorArg1, $constructorArg2]);
$mock = \Mockery::mock('MyClass', 'MyInterface', [$constructorArg1, $constructorArg2]);
```

will generate phpstan error

```
 463    Parameter #2 ...$args of static method Mockery::mock() expects class-string<Directive>|Directive, array<int, StreamFactory> given.
  716    Parameter #2 ...$args of static method Mockery::mock() expects class-string<Directive>|Directive, array<int, StreamFactory> given.
  1142   Parameter #2 ...$args of static method Mockery::mock() expects class-string<Directive>|Directive, array<int, StreamFactory> given.
  1236   Parameter #2 ...$args of static method Mockery::mock() expects class-string<Directive>|Directive, array<int, StreamFactory&Mockery\MockInterface> given.    
  1273   Parameter #2 ...$args of static method Mockery::mock() expects class-string<Directive>|Directive, array<int, StreamFactory&Mockery\MockInterface> given.    
```

Maybe something like will be better (but seems it is a breaking change and will break `phpstan/phpstan-mockery`) 

```
/**
 * @template TMock
 *
 * @param array<class-string<TMock>|TMock> $class
 *
 * @return LegacyMockInterface&MockInterface&TMock
 */
public static function mock(mixed $class, mixed ...$args)
```

Also, the `Mockery::spy(static function () {})` now gives (not sure related to `phpstan/phpstan-mockery` or not):

```
Return type of call to static method Mockery::spy() contains unresolvable type.
```

**PS:** Original issue (#1395) probably can be solved with `phpstan/phpstan-mockery` without changing docblock (at least I have used it for a while and never see any issue)

### Steps to Reproduce

Update to 1.6.10 & run phpstan

### Expected Behavior

No new phpstan errors

### Actual Behavior

A lot of new phpstan errors

**Comment by LastDragon-ru** [comment]
Seems this will solve the issue for `mock()` (number of `?mixed` determines max count of args):

```
@param array{
  <class-string<TMock>|TMock>, 
  ?mixed, 
  ?mixed, 
  ?mixed, 
  ?mixed, 
  ?mixed, 
  ?mixed, 
  ?mixed, 
  ?mixed, 
  ?mixed
} $args
```

**Comment by LastDragon-ru** [comment]
Or even better `array{<class-string<TMock>|TMock>}&array<array-key, mixed>`

**Comment by LastDragon-ru** [comment]
```php
<?php declare(strict_types = 1);

$a = Mockery::mock(stdClass::class);
$b = Mockery::mock(stdClass::class, Countable::class);
$c = Mockery::mock(stdClass::class, ['a']);
$d = Mockery::mock(stdClass::class, Countable::class, ['a']);
$e = Mockery::mock();

PHPStan\dumpType($a);
PHPStan\dumpType($b);
PHPStan\dumpType($c);
PHPStan\dumpType($d);
PHPStan\dumpType($e);

```

<details>
<summary>Previous attemps</summary>

> `array{<class-string<TMock>|TMock>, ?mixed, ?mixed, ?mixed, ?mixed, ?mixed, ?mixed, ?mixed, ?mixed, ?mixed, ?mixed}`

Without `phpstan/phpstan-mockery`:

```
 ------ ------------------------------------ 
  Line   MockeryTest.php                     
 ------ ------------------------------------
  9      Dumped type: Mockery\MockInterface
  10     Dumped type: Mockery\MockInterface
  11     Dumped type: Mockery\MockInterface
  12     Dumped type: Mockery\MockInterface
  13     Dumped type: Mockery\MockInterface
 ------ ------------------------------------
```

With

```
 ------ ------------------------------------------------------- 
  Line   MockeryTest.php                                        
 ------ -------------------------------------------------------
  9      Dumped type: Mockery\MockInterface&stdClass
  10     Dumped type: Countable&Mockery\MockInterface&stdClass
  11     Dumped type: Mockery\MockInterface&stdClass
  12     Dumped type: Countable&Mockery\MockInterface&stdClass
  13     Dumped type: Mockery\MockInterface
 ------ -------------------------------------------------------
```


> Or even better array{<class-string<TMock>|TMock>}&array<array-key, mixed>


Without `phpstan/phpstan-mockery`:

```
------ ------------------------------------ 
  Line   MockeryTest.php                     
 ------ ------------------------------------
  9      Dumped type: Mockery\MockInterface
  10     Dumped type: Mockery\MockInterface
  11     Dumped type: Mockery\MockInterface
  12     Dumped type: Mockery\MockInterface
  13     Dumped type: Mockery\MockInterface
 ------ ------------------------------------
```

With:

```
 ------ ------------------------------------------------------- 
  Line   MockeryTest.php                                        
 ------ -------------------------------------------------------
  9      Dumped type: Mockery\MockInterface&stdClass
  10     Dumped type: Countable&Mockery\MockInterface&stdClass
  11     Dumped type: Mockery\MockInterface&stdClass
  12     Dumped type: Countable&Mockery\MockInterface&stdClass
  13     Dumped type: Mockery\MockInterface
 ------ -------------------------------------------------------
```

> `@param array<class-string<TMock>|TMock> $args`

Without `phpstan/phpstan-mockery`:

```
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   MockeryTest.php                                                                                                                                
 ------ -----------------------------------------------------------------------------------------------------------------------------------------------
  5      Parameter #2 ...$args of static method Mockery::mock() expects class-string<stdClass>|stdClass, array<int, string> given.
  6      Parameter #3 ...$args of static method Mockery::mock() expects class-string<Countable|stdClass>|Countable|stdClass, array<int, string> given.
  7      Unable to resolve the template type TMock in call to method static method Mockery::mock()
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unable-to-resolve-template-type
  9      Dumped type: Mockery\MockInterface&stdClass
  10     Dumped type: (Countable&Mockery\MockInterface)|(Mockery\MockInterface&stdClass)
  11     Dumped type: Mockery\MockInterface&stdClass
  12     Dumped type: (Countable&Mockery\MockInterface)|(Mockery\MockInterface&stdClass)
  13     Dumped type: Mockery\MockInterface
 ------ -----------------------------------------------------------------------------------------------------------------------------------------------
```

With

```
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   MockeryTest.php                                                                                                                                
 ------ -----------------------------------------------------------------------------------------------------------------------------------------------
  5      Parameter #2 ...$args of static method Mockery::mock() expects class-string<stdClass>|stdClass, array<int, string> given.
  6      Parameter #3 ...$args of static method Mockery::mock() expects class-string<Countable|stdClass>|Countable|stdClass, array<int, string> given.
  7      Unable to resolve the template type TMock in call to method static method Mockery::mock()
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unable-to-resolve-template-type
  9      Dumped type: Mockery\MockInterface&stdClass
  10     Dumped type: Countable&Mockery\MockInterface&stdClass
  11     Dumped type: Mockery\MockInterface&stdClass
  12     Dumped type: Countable&Mockery\MockInterface&stdClass
  13     Dumped type: Mockery\MockInterface
 ------ -----------------------------------------------------------------------------------------------------------------------------------------------
```

</details>


> `array<class-string<TMock>|TMock|array<mixed>>`

Without `phpstan/phpstan-mockery`:

```
 ------ ------------------------------------------------------------------------------------------- 
  Line   MockeryTest.php                                                                            
 ------ -------------------------------------------------------------------------------------------
  7      Unable to resolve the template type TMock in call to method static method Mockery::mock()
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unable-to-resolve-template-type
  9      Dumped type: Mockery\MockInterface&stdClass
  10     Dumped type: (Countable&Mockery\MockInterface)|(Mockery\MockInterface&stdClass)
  11     Dumped type: Mockery\MockInterface&stdClass
  12     Dumped type: (Countable&Mockery\MockInterface)|(Mockery\MockInterface&stdClass)
  13     Dumped type: Mockery\MockInterface
 ------ -------------------------------------------------------------------------------------------

```

With:

```
 ------ ------------------------------------------------------------------------------------------- 
  Line   MockeryTest.php                                                                            
 ------ -------------------------------------------------------------------------------------------
  7      Unable to resolve the template type TMock in call to method static method Mockery::mock()
         💡 See: https://phpstan.org/blog/solving-phpstan-error-unable-to-resolve-template-type
  9      Dumped type: Mockery\MockInterface&stdClass
  10     Dumped type: Countable&Mockery\MockInterface&stdClass
  11     Dumped type: Mockery\MockInterface&stdClass
  12     Dumped type: Countable&Mockery\MockInterface&stdClass
  13     Dumped type: Mockery\MockInterface
 ------ -------------------------------------------------------------------------------------------

```

**Comment by martinssipenko** [comment]
@ghostwriter any chance this could be released as 1.6.11?

**Comment by ghostwriter** [comment]
@martinssipenko yes, `1.6.11` has been tagged.

**Comment by rolandsusans** [comment]
@ghostwriter looks like the case of arg type `Closure` was not fixed in `1.6.11`. Anyways, here is the PR, that addresses that - https://github.com/mockery/mockery/pull/1410.
